cmd/query/app/querysvc/adjuster/hash.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package adjuster
+
+import (
+	"fmt"
+	"hash/fnv"
+
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/internal/jptrace"
+)
+
+var _ Adjuster = (*SpanHashDeduper)(nil)
+
+// SpanHash creates an adjuster that deduplicates spans by removing all but one span
+// with the same hash code. This is particularly useful for scenarios where spans
+// may be duplicated during archival, such as with ElasticSearch archival.
+//
+// The hash code is generated by serializing the span into protobuf bytes and applying
+// the FNV hashing algorithm to the serialized data.
+//
+// To ensure consistent hash codes, this adjuster should be executed after
+// SortAttributesAndEvents, which normalizes the order of collections within the span.
+func SpanHash() SpanHashDeduper {
+	return SpanHashDeduper{
+		marshaler: &ptrace.ProtoMarshaler{},
+	}
+}
+
+type SpanHashDeduper struct {
+	marshaler ptrace.Marshaler
+}
+
+func (s *SpanHashDeduper) Adjust(traces ptrace.Traces) {
+	spansByHash := make(map[uint64]ptrace.Span)
+	resourceSpans := traces.ResourceSpans()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		hashTrace := ptrace.NewTraces()
+		rs := resourceSpans.At(i)
+		hashResourceSpan := hashTrace.ResourceSpans().AppendEmpty()
+		rs.Resource().Attributes().CopyTo(hashResourceSpan.Resource().Attributes())
+		scopeSpans := rs.ScopeSpans()
+		hashScopeSpan := hashResourceSpan.ScopeSpans().AppendEmpty()
+		for j := 0; j < scopeSpans.Len(); j++ {
+			ss := scopeSpans.At(j)
+			ss.Scope().Attributes().CopyTo(hashScopeSpan.Scope().Attributes())
+			spans := ss.Spans()
+			newSpans := ptrace.NewSpanSlice()
+			hashSpan := hashScopeSpan.Spans().AppendEmpty()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+				span.CopyTo(hashSpan)
+				h, err := s.computeHashCode(
+					hashTrace,
+				)
+				if err != nil {
+					jptrace.AddWarning(span, fmt.Sprintf("failed to compute hash code: %v", err))
+					span.CopyTo(newSpans.AppendEmpty())
+					continue
+				}
+				if _, ok := spansByHash[h]; !ok {
+					spansByHash[h] = span
+					span.CopyTo(newSpans.AppendEmpty())
+				}
+			}
+			newSpans.CopyTo(spans)
+		}
+	}
+}
+
+func (s *SpanHashDeduper) computeHashCode(
+	hashTrace ptrace.Traces,
+) (uint64, error) {
+	b, err := s.marshaler.MarshalTraces(hashTrace)
+	if err != nil {
+		return 0, err
+	}
+	hasher := fnv.New64a()
+	hasher.Write(b) // never returns an error
```

Comment:
Ignoring the error here because the `Hash64` interface says that the writer never returns an error. 
```go
type Hash interface {
	// Write (via the embedded io.Writer interface) adds more data to the running hash.
	// It never returns an error.
	io.Writer
```